codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/protocol/aws/DeserializeMiddleware.java

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.go.codegen.protocol.aws;
+
+import static software.amazon.smithy.go.codegen.GoWriter.goTemplate;
+import static software.amazon.smithy.go.codegen.SmithyGoDependency.SMITHY_HTTP_TRANSPORT;
+import static software.amazon.smithy.go.codegen.protocol.ProtocolUtil.hasEventStream;
+import static software.amazon.smithy.go.codegen.server.protocol.JsonDeserializerGenerator.getDeserializerName;
+
+import software.amazon.smithy.go.codegen.GoStdlibTypes;
+import software.amazon.smithy.go.codegen.GoWriter;
+import software.amazon.smithy.go.codegen.SmithyGoDependency;
+import software.amazon.smithy.go.codegen.integration.ProtocolGenerator;
+import software.amazon.smithy.model.shapes.OperationShape;
+import software.amazon.smithy.model.shapes.StructureShape;
+import software.amazon.smithy.utils.MapUtils;
+
+public class DeserializeMiddleware {
+
+    public static final String SMITHY_PROTOCOL_NAME = "awsJson10";
+    protected final ProtocolGenerator.GenerationContext ctx;
+    protected final OperationShape operation;
+    protected final GoWriter writer;
+
+    protected final StructureShape input;
+    protected final StructureShape output;
+
+    private String deserialName;
+
+    public DeserializeMiddleware(
+            ProtocolGenerator.GenerationContext ctx, OperationShape operation, GoWriter writer
+    ) {
+        this.ctx = ctx;
+        this.operation = operation;
+        this.writer = writer;
+
+        this.input = ctx.getModel().expectShape(operation.getInputShape(), StructureShape.class);
+        this.output = ctx.getModel().expectShape(operation.getOutputShape(), StructureShape.class);
+    }
+
+    public static String getMiddlewareName(OperationShape operation) {
+        return "awsAwsjson10_deserializeOp" + operation.toShapeId().getName();
+    }
+
+    public GoWriter.Writable generate() {
+        deserialName = getMiddlewareName(operation);
```

Comment:
I don't see a reason to compute this at the time generate() is called - you have access to the operation shape in the constructor and the name isn't going to change, just compute it there and then it can be a `private final String` in the class.

Reply:
This was a good point, I went ahead and reflected this change in the SerializeMiddleware as well, you will probably see the diff.